lib/wibox/container/conditional.lua

```diff
@@ -0,0 +1,362 @@
+        if (not self._private.delay_error) and type(v.when) ~= "function" then
+            self._private.delay_error = true
+            gtimer.delayed_call(function()
+                self._private.delay_error = false
+                for _, v2 in pairs(self._private.templates) do
+                    if type(v2.when) ~= "function" and not self._private.default_when then
+                        assert(false, "The conditional templates require a `when` function")
+                        return
+                    end
+                end
+            end)
+        end
+    end
+
+    self._private.templates = templates
+    self:eval()
+end
+
+--- The list of conditional templates.
+-- @property selected_template
+-- @tparam table selected_template The selected template or a widget instance
+
+function conditional:get_selected_template()
+    return self._private.selected
+end
+
+--- Return all attached signals.
+--
+-- The format is a list of tables with the object as the first value and the
+-- signal name has the second.
+--
+-- @property attached_signals
+-- @param table
+
+function conditional:get_attached_signals()
+    return self._private.attached_signal
+end
+
+--- The list of global signals used to automatically call `eval()`.
+--
+-- The format is a table of table with objects as keys and signal name as values.
+--
+-- @DOC_wibox_container_conditional_global_signals_EXAMPLE@
+--
+-- @property connected_global_signals
+-- @param table
+
+function conditional:get_connected_global_signals()
+    return self._private.connected_global_signals
+end
+
+function conditional:set_connected_global_signals(signals)
+    if self._private.connected_global_signals then
+        for _, v in pairs(self._private.connected_global_signals) do
+            v[1].disconnect_signal(v[2], self._private.eval_handler)
+        end
+    end
+
+    self._private.connected_global_signals = signals
+
+    for _, v in pairs(self._private.connected_global_signals) do
+        assert(v[1].connect_signal, "The object passed to `global_signals` is invalid")
+        assert(not v[1].is_widget,
+            "Regular objects cannot be used in `global_signals`, use `attached_signals`"
+        )
+        v[1].connect_signal(v[2], self._private.eval_handler)
+    end
+end
+
+--- Controls the arguments passed to the `when` callback.
+--
+-- This is useful when extending this container to fit specific purposes.
+--
+-- All values returned by this function will be passed to the `when` callback.
+-- Note that it will be called only once for all templates.
+--
+-- The default value returns the `conditional` widget instance.
+--
+-- The callback has a single parameter (the widget instance).
+--
+-- @property argument_callback
+-- @param function
+
+function conditional:set_argument_callback(callback)
+    self._private.argument_callback = callback
+end
+
+function conditional:get_argument_callback()
+    return self._private.argument_callback
+end
```

Comment:
An implementation of `awul.screen.only_on_screen` based on this widget could add a `screen` argument to the when callback. So you could have a theoretical `screens = {foo, bar}` field in the extended widget constructor and always pass that screen.

`wibox.container.conditional` "sucks" when used directly. It was always meant to be a base class of higher level widgets. The only one so far (because it's not merged...) is the `dynamite` declarative client layout module. It uses this mechanism to add the tag and number of clients to the `when` callback.